FV3/atmos_cubed_sphere/tools/online_coarse_graining.F90

```diff
@@ -0,0 +1,267 @@
+  use mpp_domains_mod, only: domain2d, mpp_define_io_domain, mpp_define_mosaic, mpp_get_compute_domain
+  use mpp_mod, only: FATAL, input_nml_file, mpp_error, mpp_npes
+  
+  implicit none
+  private
+
+  public :: block_sum, get_fine_array_bounds, get_coarse_array_bounds, online_coarse_graining_init, weighted_block_average, MODEL_LEVEL
+
+  interface block_sum
+     module procedure block_sum_2d
+  end interface block_sum
+  
+  interface weighted_block_average
+     module procedure weighted_block_average_2d
+     module procedure weighted_block_average_3d_field_2d_weights
+  end interface weighted_block_average
+  
+  ! Global variables for the module, initialized in online_coarse_graining_init
+  type(domain2d) :: coarse_domain
+  integer :: is, ie, js, je, npz
+  integer :: is_coarse, ie_coarse, js_coarse, je_coarse
+  integer :: target_resolution
+  character(len=11) :: MODEL_LEVEL = 'model_level'
+  
+  ! Namelist parameters initialized with default values
+  integer :: coarsening_factor = 8
+  integer :: coarse_io_layout(2) = (/1, 1/)
+  character(len=64) :: coarse_graining_strategy = 'model_level'
+  logical :: do_coarse_graining = .false.
+  
+  namelist /online_coarse_graining_nml/ coarsening_factor, coarse_io_layout, &
+       coarse_graining_strategy, do_coarse_graining
+
+contains
+
+  subroutine online_coarse_graining_init(Atm)
+    type(fv_atmos_type), intent(inout) :: Atm
+
+    character(len=256) :: error_message
+    logical :: exists
+    integer :: namelist_file, error_code, iostat
+
+#ifdef INTERNAL_FILE_NML
+    read(input_nml_file, online_coarse_graining_nml, iostat=iostat)
+    error_code = check_nml_error(iostat, 'online_coarse_graining_nml')
+#else
+    namelist_file = open_namelist_file(Atm%nml_filename)
+    read(namelist_file, online_coarse_graining_nml, iostat=iostat)
+    error_code = check_nml_error(iostat, 'online_coarse_graining_nml')
+    call close_file(namelist_file)
+#endif
+
+    Atm%coarse_graining_attrs%do_coarse_graining = do_coarse_graining
+
+    if (do_coarse_graining) then
+       call compute_target_resolution(Atm, coarsening_factor, target_resolution)
+       call assert_valid_domain_layout(target_resolution, Atm%layout)
+       call define_cubic_mosaic(coarse_domain, target_resolution, target_resolution, Atm%layout)
+       call mpp_define_io_domain(coarse_domain, coarse_io_layout)
+       call mpp_get_compute_domain(coarse_domain, is_coarse, ie_coarse, js_coarse, je_coarse)
+       call get_fine_array_bounds(Atm, is, ie, js, je)
+       npz = Atm%npz
+       
+       Atm%coarse_graining_attrs%bd%is_coarse = is_coarse
+       Atm%coarse_graining_attrs%bd%ie_coarse = ie_coarse
+       Atm%coarse_graining_attrs%bd%js_coarse = js_coarse
+       Atm%coarse_graining_attrs%bd%je_coarse = je_coarse
+       Atm%coarse_graining_attrs%target_resolution = target_resolution
+       Atm%coarse_graining_attrs%domain = coarse_domain
+       Atm%coarse_graining_attrs%strategy = coarse_graining_strategy
```

Comment:
If this is called `strategy` on the attribute, should probably also make the namelist parameter `strategy` within the coarse graining namelist.

Reply:
Yes, this is cleaner.  Changed.